src/backend/distributed/transaction/transaction_recovery.c

```diff
@@ -258,6 +261,54 @@ RecoverWorkerTransactions(WorkerNode *workerNode)
 			continue;
 		}
 
+		/* Check if the transaction is created by an outer transaction from a non-main database */
+		bool outerXidIsNull = false;
+		Datum outerXidDatum = heap_getattr(heapTuple,
+										   Anum_pg_dist_transaction_outerxid,
+										   tupleDescriptor, &outerXidIsNull);
+
+		TransactionId outerXid = 0;
+		if (!outerXidIsNull)
+		{
+			FullTransactionId outerFullXid = DatumGetFullTransactionId(outerXidDatum);
+			outerXid = XidFromFullTransactionId(outerFullXid);
+		}
+
+		if (outerXid != 0)
+		{
+			bool outerXactIsInProgress = TransactionIdIsInProgress(outerXid);
+			bool outerXactDidCommit = TransactionIdDidCommit(outerXid);
+			if (outerXactIsInProgress && !outerXactDidCommit)
+			{
+				/*
+				 * The transaction is initiated from an outer transaction and the outer
+				 * transaction is not yet committed, so we should not commit either.
+				 * We remove this transaction from the pendingTransactionSet so it'll
+				 * not be aborted by the loop below.
+				 */
+				hash_search(pendingTransactionSet, transactionName, HASH_REMOVE,
+							&foundPreparedTransactionBeforeCommit);
+				continue;
+			}
+			else if (!outerXactIsInProgress && !outerXactDidCommit)
+			{
+				/*
+				 * Since outer transaction isn't in progress and did not commit we need to
+				 * abort the prepared transaction too. We do this by simply doing the same
+				 * thing we would  do for transactions that are initiated from the main
+				 * database.
+				 */
+				continue;
```

Comment:
These two cases are currently not covered by tests yet. But it is important that they do the correct thing. They are hard to test though, because they are about edge cases that can happen after the main database transaction has already successfully committed:
1. the main db transaction is committed, but the outer has not committed yet
2. the main db transaction is committed, but the outer transaction failed to commit

I think the easiest way to trigger these cases is by simulating them using the python test framework:
1. open two connections
1. in connection A we open a transaction and get its transaction id (transaction A)
2. in connection B we manually do the flow with all our new functions using the transaction id of transaction A: start_managament_command->mark_object_distributed->execute_command_on_other_nodes->COMMIT
3. And then for edge case 2 we abort transaction A (and for edge case 1 we do nothing)
4. we run recover_prepared_transactions
5. check if the changes made by connection B have been applied or not.

For some explanation of the python test suite you can take a look here: 
1. README: https://github.com/citusdata/citus/blob/main/src/test/regress/citus_tests/test/README.md
2. Example written by Emel: https://github.com/citusdata/citus/blob/main/src/test/regress/citus_tests/test/test_extension.py